src/Rec/Rec.ts

```diff
@@ -0,0 +1,467 @@
+ * If the `Rec` doesn't contain the key, returns `None`.
+ *
+ * Uses the given equality comparer if passed. Otherwise, defaults to reference
+ * equality (triple equals) for equality comparisons.
+ *
+ * @group Lookups
+ *
+ * @returns An `Option` containing a tuple of the key and value.
+ */
+export const findWithKey =
+    <K extends string>(
+        key: NoInfer<K>,
+        { equals }: EqualityComparer<string> = EqualityComparer.Default
+    ) =>
+    <V>(rec: Rec<K, V>): Option<readonly [K, V]> => {
+        if (isEmpty(rec)) {
+            return Option.none
+        }
+
+        for (const k in rec) {
+            if (equals(k, key.toString())) {
+                return Option.some([k, rec[k]])
+            }
+        }
+
+        return Option.none
+    }
+
+/**
+ * Test whether a `Rec` contains the given key. Uses the given `EqualityComparer`
+ * if passed. Otherwise, defaults to reference equality (triple equals).
+ *
+ * @group Lookups
+ *
+ * @returns `true` if the key is in the `Rec`, `false` otherwise.
+ */
+export const containsKey =
+    <K extends string>(
+        key: NoInfer<K>,
+        equalityComparer?: EqualityComparer<string>
+    ) =>
+    <V>(rec: Rec<K, V>): boolean =>
+        pipe(rec, findWithKey(key, equalityComparer), Option.isSome)
+
+/**
+ * Get a value associated with the given key from the `Rec`. Returns a `Some`
+ * containing the value, or `None` if the key is not in the `Rec`.
+ *
+ * Uses the given equality comparer if passed, otherwise defaults to using
+ * reference equality (triple equals) for equality comparison.
+ *
+ * @group Lookups
+ */
+export const find =
+    <K extends string>(
+        key: NoInfer<K>,
+        equalityComparer?: EqualityComparer<string>
+    ) =>
+    <V extends NonNullish>(rec: Rec<K, V>): Option<V> =>
+        pipe(
+            rec,
+            findWithKey(key, equalityComparer),
+            Option.map(([, v]) => v)
+        )
+
+/**
+ * Adds a key/value pair to a `Rec`. If the given key already exists
+ * the value at that key will be updated with the given value.
+ *
+ * Will use the equality comparer if given, otherwise defaults to using
+ * reference equality (triple equals) for equality comparisons.
+ *
+ * @group Transformations
+ *
+ * @returns A new `Rec` with the added key/value pair
+ */
+export const set =
+    <K extends string, V>(
+        [key, value]: readonly [NoInfer<K>, NoInfer<V>],
+        equalityComparer?: EqualityComparer<string>
+    ) =>
+    (rec: Rec<K, V>): Rec<K, V> => {
+        if (isEmpty(rec)) {
+            return { [key]: value } as Rec<K, V>
+        }
+
+        return pipe(
+            rec,
+            findWithKey(key, equalityComparer),
+            Option.match({
+                none: () => ({
+                    ...rec,
+                    [key]: value,
+                }),
+                some: ([k]) => ({
+                    ...rec,
+                    [k]: value,
+                }),
```

Comment:
Not sure I'm following this. Won't `key` and `k` be the same in the `some` case? i.e. why can't we just `return { ...rec, [key]: value }`?

Reply:
Good question. The difference here is behavior with a custom equality comparer. I'm not sure which option makes the most sense...

For example, assume that I am using a case-insensitive string equality comparer. If I call `set(['Cheese', 100])` and there is already a key called `'cheese'`, what should happen? Do I overwrite the Key AND the Value? Or because `'cheese'` and `'Cheese'` are value-equivalent given my equality comparer, do I preserve the original Key value, since that matches the expected behavior of just overriding the value at a given key if I `set` over top of an existing one?

Which do you think is most intuitive?

Reply:
Makes total sense 👍 We should keep it as is.